### ISSUE TITLE
spdlog support for working with external logrotate

### ISSUE THREAD

**Issue by MurkyWaters95** [gabime/spdlog#3464]
We use spdlog for many C++ applications, but our environment has a mix of both python and C++ apps. I have a desire to normalize our log rotation to use the external logrotate application (https://linux.die.net/man/8/logrotate) for all our applications. Logrotate works in one of two ways, it can issue a signal to the running application, such as SIGHUP, to tell the application to re-open its log files. The other option is to use the "copytruncate" option, which copies the log file to its rotated name and then truncates the log files externally.

I mostly have interest in supporting the "copytruncate" functionality, I am wondering how much effort it would be to support the underlying file being truncated. I understand that it likely would have some performance impact, as I believe every write operation would need to do a seek to the end of the file before it writes it out. Has anyone considered implementing this feature?

For the signal option, is there an operation I can run on the various sinks to have them re-open their target files? Or do I have to fully re-construct them? Related to that, if I do reconstruct them, I would lose whatever the current logs that are buffered, correct? Since this happens after the rotation actually occurs, I don't know if I can flush the data when I get the signal or if that would just disappear into the ether. I am thinking that the file descriptor would still be valid (but in a different location) so it would flush to a file, but it would be the rotated file, meaning our external log scraping wouldn't pick up on it.

Also open to any advice in general relating to this. It would be very beneficial if an external application like logrotate could be supported, because it simplifies things for our Operations teams. We definitely cannot go through the dev effort to replace our logging in our other applications with something else, it would be feasible to update our C++ applications because the majority of them share a single library that abstracts this away.

I appreciate any thoughts you have.

Thanks,
Mike Waters

**Comment by tt4g** [comment]
> I mostly have interest in supporting the "copytruncate" functionality ...

I didn't quite understand the premise.
Does this mean you want the logger to seek to the end of the log file each time because `logrotate` truncates the log file using `copytruncate`?

The seek operation for files is handled by the OS and file system (disk driver), so the cost is unknown.
However, at least when a file is opened in append mode, the seek operation to the end of the file data is performed automatically with each write operation, so I don't think it's a cost worth worrying about.

See: https://unix.stackexchange.com/a/475621

> For the signal option, is there an operation I can run on the various sinks to have them re-open their target files? ...

Reopening is unnecessary If writing to the log file `O_APPEND` mode and copytruncate rotation.
If you want to avoid the risk of logs being lost, call `sink->flush()`: https://github.com/gabime/spdlog/wiki/Flush-policy

> Also open to any advice in general relating to this.

It appears you are attempting to control the sink in the thread that received the signal in C/C++, similar to Python's signal handlers. This is not a safe operation and may cause the program to crash or result in lost log data. I recommend thoroughly understanding the signal specification in C/C++ before starting your work.

See also: https://github.com/gabime/spdlog/issues/1607#issuecomment-654722474

**Comment by MurkyWaters95** [comment]
Thank you for the fast reply. I think I missed some details that were in my head that I didn't write down. Apologies, let me see if I can explain myself better. I also forgot to explain my attempts to do this already.

I have tried using `copytruncate` with a spdlog application already, with a `basic_file_sink_mt`. Every once in a while, I would run the application and logs would stop being written entirely after the `logrotate` application ran. So `logrotate` would correctly do the copy of the current state and then it would truncate the file, but no logs would get written to the file after the truncation. So the new logs seemed to "disappear". It did not happen every time, but seemed to happen every once in a while, so I gave up on the idea at the time and pursued a different approach with the startup script we run. I also could have tried the built in `daily_logger_mt` sink, but chose not to.

This was close to a year ago, it is possible this issue has been fixed. I can attempt to reproduce again but I figured that this was an issue with how `spdlog` writes data. Maybe that is a poor assumption, I do not understand the internals of the library at all. So maybe I need to back up and get a (semi) reproduce-able setup for triggering this behavior.

I was unaware of the `O_APPEND` mode, that is exactly what I want. Is that what `spdlog` does already or do I need to implement my own sink to control writing to a file I open myself? If `spdlog` already opens the file in `O_APPEND` mode, then I believe the majority of the functionality I want is already there.

I have not done any dev to do signal handling for this so far, wanted to gather ideas before I did anything. Your advice is appreciated on what operations are supported in a signal_handler.

I appreciate your time on this.

Thanks,
Mike Waters

**Comment by gabime** [comment]
As a quick workaround, you could try to catch SIGHUP, and set bool atomic flag `should_reopen_log=true` (like @tt4g  said, signal handling and threads don't play well together. do not try to reopen the file directly from the sighandler or almost anything else) and have dedicated thread that check periodically this flag and asks all  sinks to reopen once this flag is set.

Might work although I am not sure if some log entries won’t get lost due to race between logrotate and spdlog(logrotate truncating the log at the exact same moment spdlog writes a log entry). 

I asked chatgpt and it recommends avoiding copytruncate to and use rename+create+SIGHUP to avoid races:
```
# /etc/logrotate.d/yourapp  (no copytruncate)
# rotate by rename, create a new file, then SIGHUP the app to reopen
/var/log/yourapp/app.log {
    daily
    rotate 7
    missingok
    notifempty
    compress
    create 0644 youruser yourgroup
    postrotate
        if [ -f /run/yourapp.pid ]; then
            kill -HUP "$(cat /run/yourapp.pid)" || true
        fi
    endscript
}
```

Minimal standalone sample(from chatgpt , i didn’t try it. test thoroughly to make sure it works without losing logs): 

```c++
// sighup_spdlog_minimal.cpp
// Build: g++ -std=c++17 -O2 sighup_spdlog_minimal.cpp -pthread -o yourapp
// Run:   ./yourapp /var/log/yourapp/app.log /run/yourapp.pid

#include <spdlog/spdlog.h>
#include <spdlog/sinks/basic_file_sink.h>

#include <atomic>
#include <chrono>
#include <csignal>
#include <fstream>
#include <memory>
#include <thread>

static std::shared_ptr<spdlog::sinks::basic_file_sink_mt> g_file_sink;
static std::atomic<bool> g_hup{false};

static void hup_handler(int) noexcept { g_hup.store(true, std::memory_order_relaxed); }

static void install_sighup_handler() {
    struct sigaction sa{};
    sa.sa_handler = &hup_handler;
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = 0;
    sigaction(SIGHUP, &sa, nullptr);
}

static void write_pidfile(const char* path) {
    std::ofstream f(path);
    if (f) f << ::getpid() << "\n";
}

static void hup_watcher() {
    using namespace std::chrono_literals;
    for (;;) {
        if (g_hup.exchange(false, std::memory_order_relaxed)) {
            // flush current file, then reopen to the new file created by logrotate
            g_file_sink->flush();
            g_file_sink->truncate();
            spdlog::info("reopened after SIGHUP");
        }
        std::this_thread::sleep_for(100ms);
    }
}

int main(int argc, char** argv) {
    const char* log_path = (argc > 1) ? argv[1] : "app.log";
    const char* pid_path = (argc > 2) ? argv[2] : "app.pid";

    g_file_sink = std::make_shared<spdlog::sinks::basic_file_sink_mt>(log_path, true);
    auto logger   = std::make_shared<spdlog::logger>("app", g_file_sink);

    write_pidfile(pid_path);
    install_sighup_handler();

    std::thread(hup_watcher).detach();

    using namespace std::chrono_literals;
    for (uint64_t i = 0;; ++i) {
        logger->info("tick {}", i);  // write in infinite loop to the file sink
        std::this_thread::sleep_for(250ms);
    }
}
```

**Comment by tt4g** [comment]
> I have tried using copytruncate with a spdlog application already ...

I believe you did not set a signal handler when you tried it in the past. If you do not set a signal handler, the system's signal handler will be used. There is no guarantee that the application will work after the system's signal handler is invoked.

> If spdlog already opens the file in `O_APPEND` mode

The spdlog sink specifies the `"a"` mode character (equivalent to `O_APPEND`) when opening the log file.

https://github.com/gabime/spdlog/blob/f1d748e5e3edfa4b1778edea003bac94781bc7b7/include/spdlog/details/file_helper-inl.h#L26-L55
